File: examples/ruby/dogstatsd.rb
#!/usr/bin/env ruby

# Example: Send DogStatsD metrics from Ruby using libagent
#
# Install dependencies:
#   gem install ffi
#
# Run:
#   DYLD_LIBRARY_PATH=../../target/debug ruby dogstatsd.rb

require 'ffi'

module LibAgent
  extend FFI::Library
  
  # Determine library extension
  lib_ext = case RbConfig::CONFIG['host_os']
  when /darwin/
    'dylib'
  when /linux/
    'so'
  when /mswin|mingw/
    'dll'
  else
    'so'
  end
  
  lib_path = File.join(__dir__, '../../target/debug', "liblibagent.#{lib_ext}")
  ffi_lib lib_path

  # FFI function definitions
  attach_function :Initialize, [], :void
  attach_function :Stop, [], :void
  attach_function :SendDogStatsDMetric, [:pointer, :size_t], :int32

  class MetricsData < FFI::Struct
    layout :agent_spawns, :uint64,
           :trace_agent_spawns, :uint64,
           :agent_failures, :uint64,
           :trace_agent_failures, :uint64,
           :uptime_seconds, :double,
           :proxy_get_requests, :uint64,
           :proxy_post_requests, :uint64,
           :proxy_put_requests, :uint64,
           :proxy_delete_requests, :uint64,
           :proxy_patch_requests, :uint64,
           :proxy_head_requests, :uint64,
           :proxy_options_requests, :uint64,
           :proxy_other_requests, :uint64,
           :proxy_2xx_responses, :uint64,
           :proxy_3xx_responses, :uint64,
           :proxy_4xx_responses, :uint64,
           :proxy_5xx_responses, :uint64,
           :response_time_ema_all, :double,
           :response_time_ema_2xx, :double,
           :response_time_ema_4xx, :double,
           :response_time_ema_5xx, :double,
           :response_time_sample_count, :uint64,
           :dogstatsd_requests, :uint64,
           :dogstatsd_successes, :uint64,
           :dogstatsd_errors, :uint64
  end

  attach_function :GetMetrics, [], MetricsData.by_value
end

def send_metric(metric)
  metric_bytes = metric.encode('UTF-8')
  ptr = FFI::MemoryPointer.from_string(metric_bytes)
  LibAgent.SendDogStatsDMetric(ptr, metric_bytes.bytesize)
end

def main
  puts "DogStatsD Example - Sending Custom Metrics"
  puts "=========================================="
  puts

  # Initialize libagent
  LibAgent.Initialize
  puts "Initialized libagent"
  puts

  # Give the agent a moment to start
  sleep 1

  # Example 1: Counter - Page views
  puts "Sending counter metric (page views)..."
  result = send_metric('example.page.views:1|c|#env:prod,service:web')
  puts "✓ Counter sent: #{result == 0}"
  puts

  # Example 2: Gauge - Temperature
  puts "Sending gauge metric (temperature)..."
  result = send_metric('example.temperature:72.5|g|#location:office,sensor:A1')
  puts "✓ Gauge sent: #{result == 0}"
  puts

  # Example 3: Histogram - Response time
  puts "Sending histogram metric (response time)..."
  result = send_metric('example.request.duration:250|h|@0.5|#endpoint:/api/users')
  puts "✓ Histogram sent: #{result == 0}"
  puts

  # Example 4: Distribution - Response size
  puts "Sending distribution metric (response size)..."
  result = send_metric('example.response.size:1024|d|#status:200')
  puts "✓ Distribution sent: #{result == 0}"
  puts

  # Example 5: Timing - Database query
  puts "Sending timing metric (database query)..."
  result = send_metric('example.db.query.time:45|ms|#db:postgres,table:users')
  puts "✓ Timing sent: #{result == 0}"
  puts

  # Example 6: Set - Unique visitors
  puts "Sending set metric (unique visitors)..."
  result = send_metric('example.unique.visitors:user_12345|s')
  puts "✓ Set sent: #{result == 0}"
  puts

  # Example 7: Batched metrics
  puts "Sending batched metrics..."
  batch = <<~METRICS
    example.requests.total:1|c|#env:prod
    example.cache.hits:42|c|#cache:redis
    example.active.connections:100|g
    example.cpu.usage:75.5|g|#host:web01
  METRICS
  result = send_metric(batch.strip)
  puts "✓ Batched metrics sent (4 metrics): #{result == 0}"
  puts

  # Get metrics
  puts "Retrieving libagent metrics..."
  metrics = LibAgent.GetMetrics
  puts "DogStatsD Metrics:"
  puts "  Requests: #{metrics[:dogstatsd_requests]}"
  puts "  Successes: #{metrics[:dogstatsd_successes]}"
  puts "  Errors: #{metrics[:dogstatsd_errors]}"
  puts

  # Clean shutdown
  puts "Stopping libagent..."
  LibAgent.Stop
  puts "Done!"
end

main if __FILE__ == $PROGRAM_NAME

File: examples/java/DogStatsD.java
import com.sun.jna.*;
import com.sun.jna.ptr.PointerByReference;
import java.nio.charset.StandardCharsets;

/**
 * Example: Send DogStatsD metrics from Java using libagent
 * 
 * Compile and run:
 *   javac -cp jna-5.13.0.jar DogStatsD.java
 *   java -cp .:jna-5.13.0.jar -Djna.library.path=../../target/debug DogStatsD
 */
public class DogStatsD {

    public interface LibAgent extends Library {
        LibAgent INSTANCE = Native.load("libagent", LibAgent.class);

        void Initialize();
        void Stop();
        int SendDogStatsDMetric(byte[] payload, NativeLong payloadLen);
        MetricsData GetMetrics();
    }

    @Structure.FieldOrder({
        "agent_spawns", "trace_agent_spawns", "agent_failures", "trace_agent_failures",
        "uptime_seconds", "proxy_get_requests", "proxy_post_requests", "proxy_put_requests",
        "proxy_delete_requests", "proxy_patch_requests", "proxy_head_requests",
        "proxy_options_requests", "proxy_other_requests", "proxy_2xx_responses",
        "proxy_3xx_responses", "proxy_4xx_responses", "proxy_5xx_responses",
        "response_time_ema_all", "response_time_ema_2xx", "response_time_ema_4xx",
        "response_time_ema_5xx", "response_time_sample_count",
        "dogstatsd_requests", "dogstatsd_successes", "dogstatsd_errors"
    })
    public static class MetricsData extends Structure {
        public long agent_spawns;
        public long trace_agent_spawns;
        public long agent_failures;
        public long trace_agent_failures;
        public double uptime_seconds;
        public long proxy_get_requests;
        public long proxy_post_requests;
        public long proxy_put_requests;
        public long proxy_delete_requests;
        public long proxy_patch_requests;
        public long proxy_head_requests;
        public long proxy_options_requests;
        public long proxy_other_requests;
        public long proxy_2xx_responses;
        public long proxy_3xx_responses;
        public long proxy_4xx_responses;
        public long proxy_5xx_responses;
        public double response_time_ema_all;
        public double response_time_ema_2xx;
        public double response_time_ema_4xx;
        public double response_time_ema_5xx;
        public long response_time_sample_count;
        public long dogstatsd_requests;
        public long dogstatsd_successes;
        public long dogstatsd_errors;
    }

    private static int sendMetric(String metric) {
        byte[] metricBytes = metric.getBytes(StandardCharsets.UTF_8);
        return LibAgent.INSTANCE.SendDogStatsDMetric(metricBytes, new NativeLong(metricBytes.length));
    }

    public static void main(String[] args) throws InterruptedException {
        System.out.println("DogStatsD Example - Sending Custom Metrics");
        System.out.println("==========================================\n");

        // Initialize libagent
        LibAgent.INSTANCE.Initialize();
        System.out.println("Initialized libagent\n");

        // Give the agent a moment to start
        Thread.sleep(1000);

        // Example 1: Counter - Page views
        System.out.println("Sending counter metric (page views)...");
        int result = sendMetric("example.page.views:1|c|#env:prod,service:web");
        System.out.println("✓ Counter sent: " + (result == 0) + "\n");

        // Example 2: Gauge - Temperature
        System.out.println("Sending gauge metric (temperature)...");
        result = sendMetric("example.temperature:72.5|g|#location:office,sensor:A1");
        System.out.println("✓ Gauge sent: " + (result == 0) + "\n");

        // Example 3: Histogram - Response time
        System.out.println("Sending histogram metric (response time)...");
        result = sendMetric("example.request.duration:250|h|@0.5|#endpoint:/api/users");
        System.out.println("✓ Histogram sent: " + (result == 0) + "\n");

        // Example 4: Distribution - Response size
        System.out.println("Sending distribution metric (response size)...");
        result = sendMetric("example.response.size:1024|d|#status:200");
        System.out.println("✓ Distribution sent: " + (result == 0) + "\n");

        // Example 5: Timing - Database query
        System.out.println("Sending timing metric (database query)...");
        result = sendMetric("example.db.query.time:45|ms|#db:postgres,table:users");
        System.out.println("✓ Timing sent: " + (result == 0) + "\n");

        // Example 6: Set - Unique visitors
        System.out.println("Sending set metric (unique visitors)...");
        result = sendMetric("example.unique.visitors:user_12345|s");
        System.out.println("✓ Set sent: " + (result == 0) + "\n");

        // Example 7: Batched metrics
        System.out.println("Sending batched metrics...");
        String batch = "example.requests.total:1|c|#env:prod\n" +
                      "example.cache.hits:42|c|#cache:redis\n" +
                      "example.active.connections:100|g\n" +
                      "example.cpu.usage:75.5|g|#host:web01";
        result = sendMetric(batch);
        System.out.println("✓ Batched metrics sent (4 metrics): " + (result == 0) + "\n");

        // Get metrics
        System.out.println("Retrieving libagent metrics...");
        MetricsData metrics = LibAgent.INSTANCE.GetMetrics();
        System.out.println("DogStatsD Metrics:");
        System.out.println("  Requests: " + metrics.dogstatsd_requests);
        System.out.println("  Successes: " + metrics.dogstatsd_successes);
        System.out.println("  Errors: " + metrics.dogstatsd_errors + "\n");

        // Clean shutdown
        System.out.println("Stopping libagent...");
        LibAgent.INSTANCE.Stop();
        System.out.println("Done!");
    }
}
